src/java.base/linux/classes/jdk/internal/platform/cgroupv1/CgroupV1SubsystemController.java

```diff
@@ -44,27 +45,20 @@ public CgroupV1SubsystemController(String root, String mountPoint) {
 
     public void setPath(String cgroupPath) {
```

Comment:
Correct. I double checked the behavior in cg v2:

```
sudo cgcreate -g memory:/test
sudo cgcreate -g memory:/test/test
sudo cgset -r memory.max=50m test
sudo cgset -r memory.swap.max=0 test
sudo cgset -r memory.max=100m test/test
sudo cgset -r memory.swap.max=0 test/test
```
with the above setup, allocations smaller than the outer group size will work, bigger will be oom-killed.

```
$ sudo cgexec -g memory:/test/test sh -c "head -c 45m /dev/zero | tail | wc -c"

47185920

$ sudo cgexec -g memory:/test/test sh -c "head -c 55m /dev/zero | tail | wc -c"

0
Killed
```

I'll update the PR next week.

Reply:
When a process is moved to a subgroup (cg v1), such as with the command
```
docker run --tty=true --rm --volume=$JAVA_HOME:/jdk --privileged ubuntu:latest \
     sh -c "mkdir -p /sys/fs/cgroup/memory/test ; echo 100m > /sys/fs/cgroup/memory/test/memory.limit_in_bytes ; echo \$\$ > /sys/fs/cgroup/memory/test/cgroup.procs ; /jdk/bin/java -Xlog:os+container=trace -version"
```
the hierarchical limits may fail to be detected because of an issue in the code below
https://github.com/openjdk/jdk/blob/b9bf447209db5d7f6bb16a0310421dbe4170500c/src/hotspot/os/linux/cgroupUtil_linux.cpp#L64-L75
Here, `limit` at line 64 is not stored as a possible lowest limit, so if the inner group has lower limit than the outer group, it won't be detected (cg v2 is affected too). The cgroup path will be adjusted to the outer group (when it's limited). Another issue is in the loop at line 66 that reduces the path. In cg v1 in `cgroupns=host` mode (default) the cgroup_path includes the base cgroup and the subgroup (when moved). It may be not quite obvious which part of the path is the actual subgroup (and the CloudFoundry case has demonstrated it). It is suggested to determine the actual subgroup (path suffix) before reducing the group path. Please see the updated patch. I also added a new test case (TestMemoryWithSubgroups.java). The issue affects also Java Metrics. Maybe you could consider adding a similar fix in the path adjustment and a test case in the Metrics update ([JDK-8336881](https://bugs.openjdk.org/browse/JDK-8336881)), here's an example for Metrics test:
```
    private static void testMemoryLimitSubgroupV1(String innerSize, String outerGroupMemorySize, boolean privateNamespace) throws Exception {
        Common.logNewTestCase("testMemoryLimitSubgroup, innerSize = " + innerSize);
        DockerRunOptions opts =
            new DockerRunOptions(imageName, "sh", "-c");
        opts.javaOpts = new ArrayList<>();
        opts.appendTestJavaOptions = false;
        opts.addDockerOpts("--volume", Utils.TEST_CLASSES + ":/test-classes/")
            .addDockerOpts("--privileged")
            .addDockerOpts("--cgroupns=" + (privateNamespace ? "private" : "host"))
            .addDockerOpts("--memory", outerGroupMemorySize);
        opts.addClassOptions("mkdir -p /sys/fs/cgroup/memory/test ; " +
            "echo " + innerSize + " > /sys/fs/cgroup/memory/test/memory.limit_in_bytes ; " +
            "echo $$ > /sys/fs/cgroup/memory/test/cgroup.procs ; " +
            "/jdk/bin/java -XX:+ErrorFileToStderr -cp /test-classes/ " +
            "--add-exports java.base/jdk.internal.platform=ALL-UNNAMED " +
            "MetricsMemoryTester memory " + innerSize);

        DockerTestUtils.dockerRunJava(opts).shouldHaveExitValue(0).shouldContain("TEST PASSED!!!");
    }

    private static void testMemoryLimitSubgroupV2(String innerSize, String outerGroupMemorySize, boolean privateNamespace) throws Exception {
        Common.logNewTestCase("testMemoryLimitSubgroup, innerSize = " + innerSize);
        DockerRunOptions opts =
            new DockerRunOptions(imageName, "sh", "-c");
        opts.javaOpts = new ArrayList<>();
        opts.appendTestJavaOptions = false;
        opts.addDockerOpts("--volume", Utils.TEST_CLASSES + ":/test-classes/")
            .addDockerOpts("--privileged")
            .addDockerOpts("--cgroupns=" + (privateNamespace ? "private" : "host"))
            .addDockerOpts("--memory", outerGroupMemorySize);
        opts.addClassOptions("mkdir -p /sys/fs/cgroup/memory/test ; " +
            "echo $$ > /sys/fs/cgroup/memory/test/cgroup.procs ; " +
            "echo '+memory' > /sys/fs/cgroup/cgroup.subtree_control ; " +
            "echo '+memory' > /sys/fs/cgroup/memory/cgroup.subtree_control ; " +
            "echo " + innerSize + " > /sys/fs/cgroup/memory/test/memory.max ; " +
            "/jdk/bin/java -XX:+ErrorFileToStderr -cp /test-classes/ " +
            "--add-exports java.base/jdk.internal.platform=ALL-UNNAMED " +
            "MetricsMemoryTester memory " + innerSize);

        DockerTestUtils.dockerRunJava(opts).shouldHaveExitValue(0).shouldContain("TEST PASSED!!!");
    }
```
In cg v2 the way to enable memory controller is to move the process out of the group.

Reply:
> Here, `limit` at line 64 is not stored as a possible lowest limit, so if the inner group has lower limit than the outer group, it won't be detected (cg v2 is affected too).

Good spot! How about this to fix it?

```
jlong limit = mem->read_memory_limit_in_bytes(phys_mem); 
jlong lowest_limit = limit < 0 ? phys_mem: limit;
```

> The cgroup path will be adjusted to the outer group (when it's limited). Another issue is in the loop at line 66 that reduces the path. In cg v1 in `cgroupns=host` mode (default) the cgroup_path includes the base cgroup and the subgroup (when moved). It may be not quite obvious which part of the path is the actual subgroup (and the CloudFoundry case has demonstrated it). It is suggested to determine the actual subgroup (path suffix) before reducing the group path. Please see the updated patch.

I'm worried about the added complexity. 1.) Is this something that's needed in cg v2 too? If no, we are changing cg version agnostic code for a version specific issue. 2.) Wouldn't setting the cgroup path to `<mount_point>/<cgroup_path>` achieve the same thing, when it's currently `null` (in current master)?

After all, those are corner cases which don't seem very common.

Reply:
> Good spot! How about this to fix it?
> 
> ```
> jlong limit = mem->read_memory_limit_in_bytes(phys_mem); 
> jlong lowest_limit = limit < 0 ? phys_mem: limit;
> ```

Make sense to me.

> I'm worried about the added complexity. 1.) Is this something that's needed in cg v2 too? If no, we are changing cg version agnostic code for a version specific issue. 2.) Wouldn't setting the cgroup path to `<mount_point>/<cgroup_path>` achieve the same thing, when it's currently `null` (in current master)?

The added complexity only kicks in in cg v1 when `_root != cgroup_path`, so exactly in that corner case. In cg v2 it only works with `--cgroupns=host`, it ends up calling stat() once and then continues normally, so the added overhead is minimal.

Reply:
The added code in `CgroupUtil::adjust_controller` runs for cg v1 and cg v2 when path adjustment is deemed needed. So I'm not clear why it's needed for cg v2. I thought we have established that for cg v2 && `--cgroupns=host` the current path adjustment is sufficient? What am I missing?

Reply:
> The added code in `CgroupUtil::adjust_controller` runs for cg v1 and cg v2 when path adjustment is deemed needed. So I'm not clear why it's needed for cg v2

It looks like there's no way to see at this point, if we are in cgroup v1 or v2 - if I am not mistaken.

> I thought we have established that for cg v2 && `--cgroupns=host` the current path adjustment is sufficient? What am I missing?

The current path adjustment still needs correction of `lowest_limit` in cg v2. Also allocating `limit_cg_path` is important, that's why I added the if block.

Reply:
> It looks like there's no way to see at this point, if we are in cgroup v1 or v2 - if I am not mistaken.

On the other hand, a type parameter can be added to `CgroupUtil::adjust_controller()`. Would you recommend doing so?

Reply:
> Version specific code can be had in `set_subsystem_path()` of the corresponding impl (like an earlier version of your patch). `lowest_limit` and `limit_cg_path` fixes are version agnostic and can and should be fixed in `CgroupUtil::adjust_controller`.

Done. Please see the updated PR.

Reply:
> Done. Please see the updated PR.

The metrics part still needs the update - in the cgroup version specific `setPath()`.

Reply:
Right. I'm still not convinced this extra reduction buys us much. The adjust controller logic will handle it if kept as is in the Metrics version.

Reply:
> Right. I'm still not convinced this extra reduction buys us much. The adjust controller logic will handle it if kept as is in the Metrics version.

The adjust controller logic won't handle it, because it reduces the path from child to parent. It's goal is to locate the smallest limit.

Reply:
> > Right. I'm still not convinced this extra reduction buys us much. The adjust controller logic will handle it if kept as is in the Metrics version.
> 
> The adjust controller logic won't handle it, because it reduces the path from child to parent. It's goal is to locate the smallest limit.

It's goal is to find the path for a specific controller that has the lowest limit and keeps that path for that controller for the rest of the run-time of the JVM. Consider a case where this reduction currently happens in `set_subsystem_path()`. Instead of the manual work of setting it to a path where the directory actually exists we set it to `<mount_path>/<cgroup_path>`.

In the cloudflare case we'd end up with a subsystem path of `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c`. Since the cgroup_path != _root we trigger path adjustment increasing the chance to detect any lower limit in any of the paths down to the mount point. By doing so **and** there is a lower limit in the hierarchy we know the path exists as well and that is being used from then on.

So if  `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden` existed and has a limit it would be used. I thought this is the reason why cg v2 didn't need the same fix when `--cgroupns=host` is used.

Reply:
> In the cloudflare case we'd end up with a subsystem path of `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c`. Since the cgroup_path != _root we trigger path adjustment increasing the chance to detect any lower limit in any of the paths down to the mount point. By doing so **and** there is a lower limit in the hierarchy we know the path exists as well and that is being used from then on.

In the Cloudflare case (cg v1 before patch), the path `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c` will be adjusted as follows:
```
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/
/sys/fs/cgroup/cpu,cpuacct/system.slice/
/sys/fs/cgroup/cpu,cpuacct/
```
In cg v1 systems, all the above paths except for the very last, are invalid. So, no chance for detecting any lower limits. When a process moved to a subgroup, the trailing part of cgroup_path will constitute the subgroup path.

> So if `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden` existed and has a limit it would be used. I thought this is the reason why cg v2 didn't need the same fix when `--cgroupns=host` is used.

The correct reasoning is the following: cg v2 didn't need the same fix, because in cg v2 the entire hierarchy is mapped inside the container in --cgroupns=host mode, the entire path is accessible, including the subgroup. In cg v1 only the process's cgroup is mapped, and the mapping includes subgroups, if any.

Reply:
> In the Cloudflare case (cg v1 before patch), the path `/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c` will be adjusted as follows:

I assume that's the adjustment logic (pre-patch) that happens provided the setting to `nullptr` issue is addressed in cg v1?

> ```
> /sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c
> /sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/
> /sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/
> /sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/
> /sys/fs/cgroup/cpu,cpuacct/system.slice/
> /sys/fs/cgroup/cpu,cpuacct/
> ```
> 
> In cg v1 systems, all the above paths except for the very last, are invalid. So, no chance for detecting any lower limits. When a process moved to a subgroup, the trailing part of cgroup_path will constitute the subgroup path.

And after the patch this would become this, right?

```
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c
/sys/fs/cgroup/cpu,cpuacct/
```

If so, the change of logic isn't that different (except for where the existence checks are happening). This path adjustment only happens at boot, so the next look-up would use `/sys/fs/cgroup/cpu,cpuacct/`. So it's a trade-off, special case this use-case for added complexity or not.

Reply:
> And after the patch this would become this, right?
> 
> ```
> /sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c
> /sys/fs/cgroup/cpu,cpuacct/
> ```

It depends on whether it was a subgroup in the initial path. If `bad/2f57368b-0eda-4e52-64d8-af5c` is the subgroup, the reduction will be
```
/sys/fs/cgroup/cpu,cpuacct/system.slice/garden.service/garden/bad/2f57368b-0eda-4e52-64d8-af5c
/sys/fs/cgroup/cpu,cpuacct/bad/2f57368b-0eda-4e52-64d8-af5c
/sys/fs/cgroup/cpu,cpuacct/bad
/sys/fs/cgroup/cpu,cpuacct/
```